Start moving webapp to Redux

jwilander opened the pull request:
#### Summary
Moves the user store, preference store and part of the team store over to use Redux.

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-6212

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

enahum commented on webapp/components/channel_invite_modal.jsx:
```diff
@@ -53,7 +63,7 @@ export default class ChannelInviteModal extends React.Component {
         UserStore.addStatusesChangeListener(this.onStatusChange);
 
-        AsyncClient.getProfilesNotInChannel(this.props.channel.id, 0);
+        this.props.actions.getProfilesNotInChannel(TeamStore.getCurrentId(), this.props.channel.id, 0);
```
again you should pass the currentTeamId as a prop or use the selector here

enahum commented on webapp/components/channel_invite_modal.jsx:
```diff
@@ -103,40 +115,24 @@ export default class ChannelInviteModal extends React.Component {
 
     nextPage(page) {
-        AsyncClient.getProfilesNotInChannel(this.props.channel.id, (page + 1) * USERS_PER_PAGE, USERS_PER_PAGE);
+        this.props.actions.getProfilesNotInChannel(TeamStore.getCurrentId(), this.props.channel.id, (page + 1) * USERS_PER_PAGE, USERS_PER_PAGE);
```
same

enahum commented on webapp/components/channel_invite_modal.jsx:
```diff
@@ -103,40 +115,24 @@ export default class ChannelInviteModal extends React.Component {
-                        this.setState({search: true, users});
-                    }
-                );
+                searchUsers(term, TeamStore.getCurrentId(), {not_in_channel_id: this.props.channel.id});
```
same

enahum commented on webapp/components/login/login_controller.jsx:
```diff
@@ -209,7 +209,7 @@ export default class LoginController extends React.Component {
 
     finishSignin(team) {
-        GlobalActions.emitInitialLoad(
+        loadMe(
```
maybe im not reading this correctly but it this is called after `weblogin` that in turns calls the redux action `login` then `loadMe` is not necessary as you are going to perform the same requests more than once

jwilander replied:
Possibly, I didn't look at this one very closely

enahum commented on webapp/components/member_list_channel.jsx:
```diff
@@ -87,43 +89,30 @@ export default class MemberListChannel extends React.Component {
-                    term,
-                    TeamStore.getCurrentId(),
-                    {},
+                searchUsers(term, TeamStore.getCurrentId(), {},
```
remember this? hehe

enahum commented on webapp/components/member_list_team.jsx:
```diff
@@ -67,15 +69,15 @@ export default class MemberListTeam extends React.Component {
         }
 
-        this.setState({users: UserStore.getProfileListInTeam(), teamMembers: Object.assign([], TeamStore.getMembersInTeam())});
+        this.setState({users, teamMembers: Object.assign([], TeamStore.getMembersInTeam())});
```
selector for `membersInTeam`

enahum commented on webapp/components/more_direct_channels.jsx:
```diff
@@ -63,9 +75,9 @@ export default class MoreDirectChannels extends React.Component {
+            this.props.actions.getProfiles(0, USERS_PER_PAGE * 2);
         } else {
-            AsyncClient.getProfilesInTeam(TeamStore.getCurrentId(), 0, USERS_PER_PAGE * 2);
+            this.props.actions.getProfilesInTeam(TeamStore.getCurrentId(), 0, USERS_PER_PAGE * 2);
```
currentTeamId as a prop?

enahum commented on webapp/containers/add_users_to_team.js:
```diff
@@ -0,0 +1,26 @@
+
+import AddUsersToTeam from 'components/add_users_to_team.jsx';
+
+function makeMapStateToProps() {
```
why are we wrapping the mapStateToProps function in another function here? it doesn't seen necessary

jwilander replied:
I just copied what the RN files were doing such as https://github.com/mattermost/mattermost-mobile/blob/master/app/components/post/index.js#L19 but I guess I copied the wrong one, since that needs it for the factory selector. I'll switch them over

enahum replied:
did you forget about these?

jwilander replied:
Hold your horses, not done yet :P

enahum commented on webapp/containers/channel_invite_modal.js:
```diff
@@ -0,0 +1,26 @@
+
+import ChannelInviteModal from 'components/channel_invite_modal.jsx';
+
+function makeMapStateToProps() {
```
same here

enahum commented on webapp/containers/more_direct_channels.js:
```diff
@@ -0,0 +1,27 @@
+
+import MoreDirectChannels from 'components/more_direct_channels.jsx';
+
+function makeMapStateToProps() {
```
and here

enahum commented on webapp/containers/popover_list_members.js:
```diff
@@ -0,0 +1,26 @@
+
+import PopoverListMembers from 'components/popover_list_members.jsx';
+
+function makeMapStateToProps() {
```
again

enahum commented on webapp/containers/team_members_dropdown.js:
```diff
@@ -0,0 +1,26 @@
+
+import TeamMembersDropdown from 'components/team_members_dropdown.jsx';
+
+function makeMapStateToProps() {
```
ok this is a pattern

enahum commented on webapp/package.json:
```diff
@@ -19,8 +19,10 @@
+    "localforage": "1.5.0",
     "marked": "mattermost/marked#8f5902fff9bad793cd6c66e0c44002c9e79e1317",
     "match-at": "0.1.0",
+    "mattermost-redux": "mattermost/mattermost-redux#webapp",
```
interesting, why not master?

jwilander replied:
Because it needs a whole bunch of stuff from my redux webapp branch :P Once that's merged I'll switch it over to master but in the mean time to get testing done it's pointed at the branch

enahum commented on webapp/store/index.js:
```diff
@@ -0,0 +1,109 @@
+
+const setTransforms = [
+    ...usersSetTransform,
+    ...teamSetTransform
```
outdated!

jwilander replied:
Not sure what you mean, is there a new way to do this now? The RN repo looks like it's doing the same way unless I'm missing something

enahum replied:
```js
const usersSetTransform = [
    'profilesInChannel',
    'profilesNotInChannel',
    'profilesInTeam'
];

const teamSetTransform = [
    'membersInTeam'
];

const channelSetTransform = [
    'channelsInTeam'
];

const setTransforms = [
    ...usersSetTransform,
    ...teamSetTransform,
    ...channelSetTransform
];
```

jwilander replied:
Ah that's not even merged into `master` yet, I'll update it in a later PR

enahum commented on webapp/store/index.js:
```diff
@@ -0,0 +1,109 @@
+
+    const offlineOptions = {
+        persist: (store, options) => {
+            const persistor = persistStore(store, {storage: localForage, ...options}, () => {
```
are we going to persists the store in localstorage?

jwilander replied:
Nope, using localForage we'll be using IndexedDB or WebSQL

enahum commented on webapp/stores/redux_store.jsx:
```diff
@@ -0,0 +1,19 @@
+// This is a temporary store while we are transitioning from Flux to Redux. This file exports
+// the configured Redux store for use by actions and selectors.
+
+import configureStore from 'store';
```
im guessing this is the same as `./store` right?

jwilander replied:
Not sure what you mean, it's pulling in webapp/store/index.js with this

jwilander commented on webapp/actions/user_actions.jsx:
```diff
@@ -19,6 +19,64 @@ import Client from 'client/web_client.jsx';
+                });
+            }
+
+            localStorage.setItem('currentUserId', UserStore.getCurrentId());
```
This file will get deleted eventually anyways so it doesn't really matter

jwilander commented on webapp/actions/user_actions.jsx:
```diff
@@ -19,6 +19,64 @@ import Client from 'client/web_client.jsx';
+
+            localStorage.setItem('currentUserId', UserStore.getCurrentId());
+
+            getMyTeams()(dispatch, getState).then(() => {
```
Hmmm, for some reason I thought it wasn't. I'll remove this

jwilander commented on webapp/actions/user_actions.jsx:
```diff
@@ -491,119 +470,69 @@ function onThemeSaved(teamId, theme, onSuccess) {
-        channelId,
+export function autocompleteUsersInChannel(username, channelId, success) {
+    const channel = ChannelStore.get(channelId);
+    const teamId = channel ? channel.team_id : TeamStore.getCurrentId();
```
Like I said, doesn't really matter as these actions will get moved into redux directly

jwilander commented on webapp/actions/user_actions.jsx:
```diff
@@ -907,7 +794,7 @@ export function uploadProfileImage(userPicture, success, error) {
         userPicture,
         () => {
-            AsyncClient.getMe();
+            getMe()(dispatch, getState);
```
Are you thinking of loadMe?

jwilander commented:
Regarding the majority of your comments:

No point in using the selectors over the flux stores in the action files, all those actions will be moved to the Redux repo anyways. As for the props vs using the selectors, those changes will come later when those components actually become pure.

mattermod commented:
Spinmint test server created at: http://i-098ca493ab8411fa4.spinmint.com:8065/pr6140

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-098ca493ab8411fa4

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-06816894d507db618.spinmint.com:8065/pr6140

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-06816894d507db618

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0181d3e68f71b68ac.spinmint.com:8065/pr6140

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0181d3e68f71b68ac

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0529fd2bb39cfcd6b.spinmint.com:8065/pr6140

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0529fd2bb39cfcd6b

jasonblais commented:
@jwilander Tested the following. I can continue testing more later, but I think I found a few things that should keep you busy for a bit :)

 - promoting member to team admin, channel admin
 - posting messages, replies, edits, permalink view, search, recent mentions, pinned/flagged posts searches
 - pinning posts, flagging, 
 - checking statuses
 - adding users to a team and channel
 - creating DM/GM channels
 - updating account settings > general info
 - viewing/revoking sessions
 - clicking/checking random things on the Account Settings modal

1) Main Menu is missing options for a regular member and team admins

Observed: Missing fields
 - get team invite link [this seems to reproduce on `master` so I can create a separate ticket for this]
 - create team [team creation is enabled]
 - join another team
 - report a problem / help / download apps links

Note that these options are shown for a system admin. None of these options have been restricted to admins only from the System Console

2) After a refresh, all statuses appear offline.

![image](https://cloud.githubusercontent.com/assets/13119842/25208151/abfa207c-2540-11e7-9cb5-1e2a625ec2ad.png)
^haven't used slash commands to manually set status. All accounts are online / had activity in the last 60 seconds.

3) Not receiving any desktop notifications in Chrome (only hear a 'ping' sound) nor in Firefox

Notification settings:
![image](https://cloud.githubusercontent.com/assets/13119842/25208185/f1163fec-2540-11e7-8807-031ef55d33e7.png)

4) JS error after reloading a channel (or switching to a channel) containing messages

![image](https://cloud.githubusercontent.com/assets/13119842/25208205/181fa9de-2541-11e7-9a4e-2c5c90baa4d2.png)

5) Replies don't work (hitting Enter in RHS does nothing + get JS errors)

![image](https://cloud.githubusercontent.com/assets/13119842/25208216/32a60424-2541-11e7-82ef-7d56cbf878e6.png)

6) JS error each time after typing a letter following an "@"

![image](https://cloud.githubusercontent.com/assets/13119842/25208271/81efb2f0-2541-11e7-8719-05df6a4b98e5.png)

7) Recent mentions don't work [clicking the "@" sign does nothing]

8) Flagging a post doesn't work [clicking the flag icon does nothing]

9) Cannot open a DM with a person outside the team

 - go to System Console > Users and Teams and set "Enable users to open Direct Message channels with:" to "Any user on the Mattermost server"
![image](https://cloud.githubusercontent.com/assets/13119842/25208325/f71c34c2-2541-11e7-9a22-9f86d758de57.png)
 - create a team
 - click "More..." at the bottom of LHS to open "More DM" modal
 - observed: No one to choose from [but switching to a team with members let's me start a DM or GM with those people]
![image](https://cloud.githubusercontent.com/assets/13119842/25208364/39ce8aea-2542-11e7-84af-b365b0cf776f.png)

10) Random: I saw this exception in the console sometimes

![image](https://cloud.githubusercontent.com/assets/13119842/25208387/5dc6e8fc-2542-11e7-8e89-99ba07957b9c.png)

11) viewing Account Settings > Security > Active Sessions opens a modal, but info never loads
![image](https://cloud.githubusercontent.com/assets/13119842/25208440/c1ecebba-2542-11e7-85e0-ff2e5673310a.png)

12) viewing active sessions from Account Settings > Security opens a "help" dialog. But can't actually view or revoke them

![image](https://cloud.githubusercontent.com/assets/13119842/25208482/062b5f3c-2543-11e7-81cb-7f16f965b079.png)

13) When mobile push notifications aren't configured/enabled, help text for mobile push notifications should say "Disabled by System Administrator"

Observed [expanding the settings does say it's disabled; but it doesn't say so in the attached help text]:
![image](https://cloud.githubusercontent.com/assets/13119842/25208522/426375de-2543-11e7-9c9d-d683e533d19d.png)
![image](https://cloud.githubusercontent.com/assets/13119842/25208530/52423cc4-2543-11e7-9ebb-5fe0bcaa1e46.png)
^second screenshot is correct


Expected:
![image](https://cloud.githubusercontent.com/assets/13119842/25208503/2712a1c4-2543-11e7-8137-fe753b99c07b.png)

14) Selecting options in Account Settings > Notifications don't get saved (after saving, resets to default option)

15) Updating password gives "Invalid Date" as the time of last update [the password does update though]
![image](https://cloud.githubusercontent.com/assets/13119842/25208631/e54f188e-2543-11e7-9c59-062f2b21119f.png)

16) Updating themes doesn't work (cannot save changes + JS error)

![image](https://cloud.githubusercontent.com/assets/13119842/25208650/1077fbac-2544-11e7-8e54-15579fa6b972.png)

17) Updating font doesn't work
![image](https://cloud.githubusercontent.com/assets/13119842/25208677/32b7bd60-2544-11e7-92a2-70b9294162e7.png)

^I think the first error came from clicking outside the Account Settings modal

jwilander commented:
@jasonblais Nice work! That was extremely helpful. Most of those issues arose from just a few cascading problems and I've pushed fixes for all of them

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0de8a49cbe3c4ea11.spinmint.com:8065/pr6140

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0de8a49cbe3c4ea11

jasonblais commented:
Tested the following on Win10 / Chrome + Firefox

 - confirmed previously reported issues are fixed
 - promoting member to team admin, channel admin
 - posting messages, replies, edits, permalink view, search, recent mentions, pinned/flagged posts searches
 - clicking "Load more messages"
 - pinning posts, flagging, 
 - checking statuses
 - adding users to a team and channel
 - creating DM/GM channels
 - updating account settings > general info
 - viewing/revoking sessions
 - clicking/checking random things on the Account Settings modal


General observation: I feel the initial load of some features is slower: E.g. my status is shown as offline for a split second before it changes to online. Also, sometimes modals load only partially until all information is displayed (again, the delay is small, but noticeable).

1) Pinning a post works but gives a JS error

![image](https://cloud.githubusercontent.com/assets/13119842/25247775/e83371c6-25d9-11e7-8364-9a729125c26f.png)

2) Recent mentions search doesn't work (clicking "@" does nothing)

3) Replying to a post in flagged posts list throws an error

 - flag a post
 - click on the flag icon to display flagged posts list
 - click reply arrow
 - post a reply
 - observed:
![image](https://cloud.githubusercontent.com/assets/13119842/25247835/08911202-25da-11e7-892e-d4ee93ab00e5.png)

4) Initially desktop notifications worked on Chrome, but then I started getting JS errors

![image](https://cloud.githubusercontent.com/assets/13119842/25248267/98f98788-25db-11e7-99f8-5ab325f4f090.png)

5) When attempting to change a password, entering an incorrect "Current Password" gives wrong error message and JS errors

![image](https://cloud.githubusercontent.com/assets/13119842/25252847/90ea2b7c-25ec-11e7-9d98-5df557a03308.png)

Expected: Error message `The "Current Password" you entered is incorrect. Please check that Caps Lock is off and try again.`

6) Revoking a session changes UI and causes a JS error before logs you out

jasonblais commented:
Argh, sorry for delay, battling connection issues :/

----

Tested the following on Win10 / Chrome + Firefox

 - confirmed previously reported issues are fixed
 - promoting member to team admin, channel admin
 - posting messages, replies, edits, permalink view, search, recent mentions, pinned/flagged posts searches
 - clicking "Load more messages"
 - pinning posts, flagging, 
 - checking statuses
 - adding users to a team and channel
 - creating DM/GM channels
 - updating account settings > general info
 - viewing/revoking sessions
 - clicking/checking random things on the Account Settings modal

I didn't test settings around System Console (e.g. user permissions) nor logging in/out of various account types, nor WebRTC yet

----

General observation: I feel the initial load of some features is slower: E.g. my status is shown as offline for a split second before it changes to online. Also, sometimes modals load only partially until all information is displayed (again, the delay is small, but noticeable).

1) Pinning a post works but gives a JS error

![image](https://cloud.githubusercontent.com/assets/13119842/25247775/e83371c6-25d9-11e7-8364-9a729125c26f.png)

2) Recent mentions search doesn't work (clicking "@" does nothing)

3) Replying to a post in flagged posts list throws an error

 - flag a post
 - click on the flag icon to display flagged posts list
 - click reply arrow
 - post a reply
 - observed:
![image](https://cloud.githubusercontent.com/assets/13119842/25247835/08911202-25da-11e7-892e-d4ee93ab00e5.png)

4) Initially desktop notifications worked on Chrome, but then I started getting JS errors.

They started working again after logout/login.

![image](https://cloud.githubusercontent.com/assets/13119842/25248267/98f98788-25db-11e7-99f8-5ab325f4f090.png)

5) When attempting to change a password, entering an incorrect "Current Password" gives wrong error message and JS errors

![image](https://cloud.githubusercontent.com/assets/13119842/25252847/90ea2b7c-25ec-11e7-9d98-5df557a03308.png)

Expected: Error message `The "Current Password" you entered is incorrect. Please check that Caps Lock is off and try again.`

6) Revoking a session changes UI and causes a JS error before logs you out

![image](https://cloud.githubusercontent.com/assets/13119842/25252992/1959d8cc-25ed-11e7-9d9a-a6294ef9131f.png)

7) Typing in a channel containing replies throws JS errors (doesn't reproduce consistently, reproduces more frequently for text containing `@` or `~` that opens an autocomplete)

![image](https://cloud.githubusercontent.com/assets/13119842/25253174/c03aa694-25ed-11e7-942e-5b6c0b981cf8.png)

8) The spinmint site doesn't load for me on iOS Safari browser (can't get to the login page, just loads forever). Not sure if it's on my device (due to connection issues), but other MM sites seemed to load fine

9) Emoji sprites on the picker aren't visible

![image](https://cloud.githubusercontent.com/assets/13119842/25254329/014c09de-25f3-11e7-9905-ff2becadaa81.png)

10)

jwilander commented:
Alright thanks @jasonblais, I think I fixed most of the issues:

1) I'm not sure what was causing this but I can't reproduce and the code in that JS error isn't related to anything I touched. Not sure if you found some corner case bug but I don't think it's related to my changes
2/3/4/7) Were all caused by the same issue, which was the the current user was getting overwritten by autocomplete (so your notification preferences would get wiped if you typed '@')
5) Fixed after I saw your accidental post ;)
8) Not sure why it's not working, going to charge my iPod to try it out
9) I didn't change anything related to the emoji picker and it works fine for me, maybe it's your connection issues?

As for the general slowness you observed, the only thing I can think of is that the initial load might be a bit slower since Redux does that as different requests. That should only affect the loading before anything at all renders in the UI though, and won't affect statuses or modals at all. For me it doesn't seem any slower but maybe something to watch out for

jwilander commented:
Figured out and fixed it on iOS Safari. The reason that didn't work is because javascript is dumb :/

jasonblais commented:
1. Oh right, this is an edge case I found earlier on `master`. Help wanted ticket here, no need to fix in this PR: https://github.com/mattermost/platform/issues/6145
9. True, might be connection-related, will re-test on the new spinmint once it's up

Thanks for the others!

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0ac2f29f4071b2d9d.spinmint.com:8065/pr6140

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0ac2f29f4071b2d9d

jwilander commented:
Hmmm seems something is broken, let me check

mattermod commented:
Spinmint test server destroyed

jwilander commented:
Ok it was a change to EN's redux-offline fork that broke it for production. I've fixed the issue and started the rebuild

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0946944d7d6e22cc3.spinmint.com:8065/pr6140

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0946944d7d6e22cc3

jwilander commented:
@jasonblais Alright, I think it's finally ready

jasonblais commented:
Confirmed above fixes are done, except 

1 - has a help wanted ticket: #6145
9 - working, probably caused by slow connection then
10 - There's a delay between revoking a session and actually getting the session revoked (~5 seconds). On `master` the action happens immediately (< 1s)



![image](https://cloud.githubusercontent.com/assets/13119842/25291315/8908e53e-269f-11e7-9d85-9a9a5e83c31c.png)

11 - I can't seem to connect to an AD/LDAP server. Tried configuring based on 
 - https://github.com/mattermost/platform-private/blob/master/config/config_pre_release.json
 - https://github.com/mattermost/platform-private/blob/master/config/config_ci_linux_mysql.json

![image](https://cloud.githubusercontent.com/assets/13119842/25292257/54d99854-26a3-11e7-88e6-6094f1747cc0.png)

WebRTC: Works on Firefox, needed a couple of tries on Chrome but I think it's related to known connection issues on WebRTC.

No other issues found so far

jwilander commented:
10 - I believe the delay is based on when your next API request goes out, so sometimes it will be immediate if you click on something that causes a request, other times it will be a few seconds until your browser makes another request

11 - The LDAP server is behind the VPN and the spinmints aren't so they won't be able to connect to it, I think this one is fine to test once it's on CI

jasonblais commented:
@jwilander Any other areas that come to mind that I should test?

There might still be some edge cases, but haven't found any new ones so far..

If we get it soaking on pre-release, we'll hopefully catch the remaining bugs before code complete + LI or LM will do some testing before closing the Jira ticket

jwilander commented:
Yeah, let's get it dev reviewed and then soak it on pre-release to find the
corner cases

On Apr 21, 2017 5:34 PM, "Jason Blais" <notifications@github.com> wrote:

> @jwilander <https://github.com/jwilander> Any other areas that come to
> mind that I should test?
>
> There might still be some edge cases, but haven't found any new ones so
> far..
>
> If we get it soaking on pre-release, we'll hopefully catch the remaining
> bugs before code complete + LI or LM will do some testing before closing
> the Jira ticket
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mattermost/platform/pull/6140#issuecomment-296310697>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACjF4mGdJImvdkjF7-VH5fi_o5XrsDz9ks5rySDNgaJpZM4M_twg>
> .
>
